var/spack/repos/builtin/packages/jdk/package.py

```diff
@@ -22,60 +22,167 @@
-    version('8u131-b11', '75b2cb2249710d822a60f83e28860053', curl_options=curl_options,
+    version('1.8.0_131-b11', '75b2cb2249710d822a60f83e28860053', curl_options=curl_options,
             url='http://download.oracle.com/otn-pub/java/jdk/8u131-b11/d54c1d3a095b4ff2b6607d096fa80163/jdk-8u131-linux-x64.tar.gz')
-    version('8u92-b14',  '65a1cc17ea362453a6e0eb4f13be76e4', curl_options=curl_options)
-    version('8u73-b02',  '1b0120970aa8bc182606a16bf848a686', curl_options=curl_options)
-    version('8u66-b17',  '88f31f3d642c3287134297b8c10e61bf', curl_options=curl_options)
-    # The 7u80 tarball is not readily available from Oracle.  If you have
-    # the tarball, add it to your mirror as mirror/jdk/jdk-7u80.tar.gz and
-    # away you go.
-    version('7u80-b0', '6152f8a7561acf795ca4701daa10a965')
-
-    provides('java@8', when='@8u0:8u999')
-    provides('java@7', when='@7u0:7u999')
-
-    def url_for_version(self, version):
-        url = "http://download.oracle.com/otn-pub/java/jdk/{0}/jdk-{1}-linux-x64.tar.gz"
-        version = str(version)
-        minor_version = version[:version.index('-')]
-        return url.format(version, minor_version)
+    version('1.8.0_92-b14', '65a1cc17ea362453a6e0eb4f13be76e4', curl_options=curl_options)
+    version('1.8.0_73-b02', '1b0120970aa8bc182606a16bf848a686', curl_options=curl_options)
+    version('1.8.0_66-b17', '88f31f3d642c3287134297b8c10e61bf', curl_options=curl_options)
+    version('1.7.0_80-b0', '6152f8a7561acf795ca4701daa10a965', curl_options=curl_options)
+
+    provides('java')
+    provides('java@10', when='@10.0:10.999')
+    provides('java@9',  when='@9.0:9.999')
+    provides('java@8',  when='@1.8.0:1.8.999')
+    provides('java@7',  when='@1.7.0:1.7.999')
+
+    # TODO: Override activate and deactivate to
+    # symlink all .jar files to <prefix>/lib/ext
+    extendable = True
+
+    @property
+    def home(self):
+        """Most of the time, ``JAVA_HOME`` is simply ``spec['java'].prefix``.
+        However, if the user is using an externally installed JDK, it may be
+        symlinked. For example, on macOS, the ``java`` executable can be found
+        in ``/usr/bin``, but ``JAVA_HOME`` is actually
+        ``/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home``.
+        Users may not know the actual installation directory and add ``/usr``
+        to their ``packages.yaml`` unknowingly. Run ``java_home`` if it exists
+        to determine exactly where it is installed. Specify which version we
+        are expecting in case multiple Java versions are installed.
+        See ``man java_home`` for more details."""
+
+        prefix = self.prefix
+        java_home = prefix.libexec.java_home
+        if os.path.exists(java_home):
+            java_home = Executable(java_home)
+            version = str(self.version.up_to(2))
+            prefix = java_home('--version', version, output=str).strip()
```

Comment:
Technically, this isn't perfect. If someone has multiple versions of JDK 10.0 installed, it will return the newest version.